File: July-LeetCoding-Challenge-2020/Week3/151-reverse-words-in-a-string.py
class Solution:
    # Sol 1. Built-in
    def reverseWords(self, s: str) -> str:
        return " ".join(reversed(s.split()))

    # TC: O(n), split runs n words in the string
    # SC: O(n), split provides n words space

    # Sol 2. Using two pointers
    def reverseWords(self, s: str) -> str:
        result,i,n = '' , 0 , len(s)
        while i<n:
            while i<n and s[i] == ' ':
                i+=1
            if i >= n: 
                break
            j=i+1
            while j<n and s[j] !=' ': 
                j+=1
                
            substring = s[i:j]
            if len(result) == 0: 
                result = substring
            else: 
                result = substring + ' ' + result
            i = j+1
        return result
    # TC: O(n)
    # SC: O(n)

File: July-LeetCoding-Challenge-2020/Week3/151-reverse-words-in-a-string.js
/**
 * @param {string} s
 * @return {string}
 * Runtime: 64 ms, faster than 83.16% of JavaScript online submissions for Reverse Words in a String.
 * Memory Usage: 37.5 MB, less than 21.29% of JavaScript online submissions for Reverse Words in a String.
 * TC: O(N), where N is length of s
 * SC: O(N), where N is length of s
 */
const SPACE = ' ';

var reverseWords = function(s) {
    const result = [];

    for (let i = s.length - 1; i >= 0; i--) {
        while (i >= 0 && s.charAt(i) === SPACE) {
            i--;
        }

        const wordEnd = i + 1;
        while (i >= 0 && s.charAt(i) !== SPACE) {
            i--;
        }

        const wordStart = i + 1;

        if (wordStart < wordEnd) {
            result.push(s.substring(wordStart, wordEnd));
        }
    }

    return result.join(SPACE);
};